### PR TITLE
Replace rusoto with custom implementation for AWS (#2176) 

### LINKED ISSUE #2350: AmazonS3 Support IMDSv2
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

IMDSv1 has [known vulnerabilities](http://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/) consequently is commonly disabled, currently rusoto does not support this - https://github.com/rusoto/rusoto/issues/1818

**Describe the solution you'd like**

Migrating away from rusoto as part of #2176 would achieve this

**Describe alternatives you've considered**

We could also move to the new AWS SDKs, although they are marked experimental, and have a huge dependency footprint

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2350
Part of #2176

# Rationale for this change
 

See #2176

# What changes are included in this PR?

Replaces rusoto with a custom implementation based on reqwest

# Are there any user-facing changes?

Not that I'm aware of, but there is a possibility

### CODE REVIEW

File: object_store/src/aws/credential.rs [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+const DATE_HEADER: &'static str = "x-amz-date";
+const HASH_HEADER: &'static str = "x-amz-content-sha256";
+const TOKEN_HEADER: &'static str = "x-amz-security-token";
+const AUTH_HEADER: &'static str = "authorization";
+
+const ALL_HEADERS: &[&'static str; 4] =
+    &[DATE_HEADER, HASH_HEADER, TOKEN_HEADER, AUTH_HEADER];
+
+impl<'a> RequestSigner<'a> {
+    fn sign(&self, request: &mut Request) {
```
Comment (tustvold) [inline object_store/src/aws/credential.rs:80]: Everything about this process is ludicrous... TLS exists... Why oh why would you do this :sob:

File: object_store/src/aws/credential.rs [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+        cache: TokenCache<Arc<AwsCredential>>,
+    },
+    WebIdentity {
+        cache: TokenCache<Arc<AwsCredential>>,
+        token: String,
+        role_arn: String,
+        session_name: String,
+        endpoint: String,
+    },
+}
```
Comment (tustvold) [inline object_store/src/aws/credential.rs:263]: I have not yet tested InstanceProvider or WebIdentityProvider, they are therefore probably slightly broken atm

File: object_store/src/aws/credential.rs [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+async fn instance_creds(
+    client: &Client,
+    retry_config: &RetryConfig,
+) -> Result<TemporaryToken<Arc<AwsCredential>>, StdError> {
+    const AWS_CREDENTIALS_PROVIDER_IP: &str = "169.254.169.254";
+    const AWS_CREDENTIALS_PROVIDER_PATH: &str =
+        "latest/meta-data/iam/security-credentials";
+    const AWS_EC2_METADATA_TOKEN_HEADER: &str = "X-aws-ec2-metadata-token";
+
+    let token_url = format!("http://{}/latest/api/token", AWS_CREDENTIALS_PROVIDER_IP);
```
Comment (tustvold) [inline object_store/src/aws/credential.rs:344]: This is the fix for #2350

File: object_store/src/multipart.rs [hunk 4]
```diff
@@ -26,23 +27,19 @@ type BoxedTryFuture<T> = Pin<Box<dyn Future<Output = Result<T, io::Error>> + Sen
 /// A trait that can be implemented by cloud-based object stores
 /// and used in combination with [`CloudMultiPartUpload`] to provide
 /// multipart upload support
-///
-/// Note: this does not use AsyncTrait as the lifetimes are difficult to manage
-pub(crate) trait CloudMultiPartUploadImpl {
+#[async_trait]
```
Comment (tustvold) [inline object_store/src/multipart.rs:30]: I took the opportunity to clean up this interface a bit, this isn't public and so this isn't a breaking change. I may split this out into another PR

File: object_store/src/multipart.rs [hunk 5]
```diff
@@ -185,10 +186,15 @@ where
 
         // If shutdown task is not set, set it
         let parts = std::mem::take(&mut self.completed_parts);
+        let parts = parts.into_iter().map(Option::unwrap).collect();
```
Comment (tustvold) [inline object_store/src/multipart.rs:189]: I need to double-check if this needs to propagate an error, I'm not sure under what circumstances a part might be missing

**Comment by tustvold** [comment]
Looks like I found a bug in the azure implementation :sweat_smile:

File: object_store/src/aws/credential.rs (hunk 1 above)
Comment (crepererum) [reply object_store/src/aws/credential.rs:79]: Big cooperation firewall/VPN stuff with TLS stripping :see_no_evil:

Comment (tustvold) [reply object_store/src/aws/credential.rs:79]: Previous comment about this being ludicrous still stands :laughing:

File: object_store/src/aws/credential.rs [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+        Self {
+            key_id: s.access_key_id,
+            secret_key: s.secret_access_key,
+            token: Some(s.session_token),
+        }
+    }
+}
+
+/// <https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html>
+async fn web_identity(
```
Comment (tustvold) [inline object_store/src/aws/credential.rs:410]: I don't have a good way to integration test this, but I have tested it with some real EKS tokens and it seems to work

File: object_store/src/aws/credential.rs [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+        let authorization_val = HeaderValue::from_str(&authorisation).unwrap();
+        request.headers_mut().insert(AUTH_HEADER, authorization_val);
+    }
+}
+
+pub trait CredentialExt {
+    fn sign(self, credential: &AwsCredential, region: &str, service: &str) -> Self;
+}
+
+impl CredentialExt for RequestBuilder {
```
Comment (crepererum) [inline object_store/src/aws/credential.rs:148]: I'm always a bit torn when it comes to extension traits for foreign types. Sure the make the code more elegant (I can clearly see that that in `client.rs`, which now feels like rather normal `reqwest` client stuff with very few parameters) but looking at code that uses is often makes me wonder "what's that new/fancy method on this type I've worked with before?!". So you can leave it as it is, just don't overdo it :wink: 

Also: they are technically prone to break under semantic versioning, because upstream may be free to add new methods (not a breaking change) which then collide with this method name. Since your "use side" doesn't use a fully qualified name, the code will stop compiling due to ambiguity.

One thing to avoid both (my confusion and the potential breakage) would be to choose a more specific method name like `aws_sign` or `s3_sign`.

File: object_store/src/aws/mod.rs [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+            store: "S3",
+            source: Box::new(err),
+        }
+    }
+}
+
+/// Interface for [Amazon S3](https://aws.amazon.com/s3/).
+#[derive(Debug)]
+pub struct AmazonS3 {
+    client: Arc<S3Client>,
```
Comment (crepererum) [inline object_store/src/aws/mod.rs:113]: What's the reason that the `S3Client` is not just inlined into the store object? Feels a bit like a layer that isn't really necessary.

File: object_store/src/aws/credential.rs [hunk 9]
```diff
@@ -0,0 +1,526 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (crepererum) [inline object_store/src/aws/credential.rs:1]: This module makes me a bit unhappy. It's such a PITA that AWS requires all that stuff. However I see you clearly did your homework and it's documented and the tests work, so I just assume it does the right thing.

If you want me to double-check the code with the official docs, I can do that but I think it's not really required.

File: object_store/src/aws/credential.rs [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+    for byte in bytes {
+        // String writing is infallible
+        let _ = write!(out, "{:02x}", byte);
+    }
+    out
+}
+
+/// Canonicalizes headers into the AWS Canonical Form.
+///
+/// <http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html>
```
Comment (crepererum) [inline object_store/src/aws/credential.rs:202]:
```suggestion
/// <https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html>
```

File: .github/workflows/object_store.yml [hunk 11]
```diff
@@ -59,6 +59,13 @@ jobs:
         image: localstack/localstack:0.14.4
         ports:
           - 4566:4566
+      ec2-metadata:
```
Comment (crepererum) [inline .github/workflows/object_store.yml:62]: Why do we need EC2 now? Is this to test that EC2 VMs could fetch S3 credentials using yet-another-mechanism[^1]?

[^1]: That's AWS blaming, it's not against this PR.

**Review by crepererum — COMMENTED** [review]
Finished the first round. Good job! I think this will definitely pay off.

File: object_store/src/aws/mod.rs (hunk 8 above)
Comment (tustvold) [reply object_store/src/aws/mod.rs:113]: It helps with encapsulation, especially as many methods on `ObjectStore` map to some combination of actual upstream requests - e.g. the various types of Put request. Also made it easier to split into its own module

File: .github/workflows/object_store.yml (hunk 11 above)
Comment (tustvold) [reply .github/workflows/object_store.yml:62]: Yup - this is the instance metadata mechanism for getting credentials

File: object_store/src/aws/mod.rs [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+    #[snafu(display("Error reading token file: {}", source))]
+    ReadTokenFile { source: std::io::Error },
+}
+
+impl From<Error> for super::Error {
+    fn from(err: Error) -> Self {
+        Self::Generic {
+            store: "S3",
+            source: Box::new(err),
+        }
```
[the remark this answers is not in the document]
Comment (tustvold) [reply object_store/src/aws/mod.rs:106]: The NotFound variant is produced by the client

**Comment by tustvold** [comment]
https://github.com/influxdata/influxdb_iox/pull/5342 has been deployed and appears to be working correctly, testing both the Instance Metadata credentials and the WebIdentity

**Comment by alamb** [comment]
I would like to request holding off merging this PR until we have done the official 0.4 object store release from this repo, tracked by https://github.com/apache/arrow-rs/issues/2338

My rationale is that there are enough moving parts and we are (effectively) still testing some of this code in production so I don't want to accidentally introduce a regression that would require a new release quickly after the previous one

File: object_store/src/aws/client.rs [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+impl S3Config {
+    fn path_url(&self, path: &Path) -> String {
+        format!("{}/{}/{}", self.endpoint, self.bucket, encode_path(path))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct S3Client {
+    config: S3Config,
+    client: Client,
```
[the remark this answers is not in the document]
Comment (alamb) [reply object_store/src/aws/client.rs:214]: image

**Comment by alamb** [comment]
Sounds like the IOx 'testing in production' has gone very well ❤️ 

So to move this PR forward:
1. It seems like like there are a few more comments from @crepererum  to address. 
2. The 0.4.0 release is going along nicely, 
3. I think we are on track to merge this over the weekend or early next week

**Comment by tustvold** [comment]
I intend to get this in this afternoon unless anybody would like longer to review

**Comment by alamb** [comment]
I don't have any objections, though I haven't reviewed this PR in detail either. I will confirm we deployed these changes in our IOx environment and they seem to be working great

cc @roeap

**Comment by alamb** [comment]
I don't have any objections, though I haven't reviewed this PR in detail either. I will confirm we deployed these changes in our IOx environment and they seem to be working great

cc @roeap

**Review by crepererum — APPROVED** [review]
